### PR TITLE
CRW-1422 Add async-sidecar productized container

### PR DESCRIPTION
**Pull request by mkuznyetsov**
Signed-off-by: Mykhailo Kuznietsov <mkuznets@redhat.com>

Add async-sidecar container, as per https://issues.redhat.com/browse/CRW-1422 issue

### CODE REVIEW

File: codeready-workspaces-devfileregistry/job-config.json [hunk 1]
```diff
@@ -2,6 +2,36 @@
   "Purpose": "To set configurations of jenkins jobs including upstream branches, dependencies and disabled status",
   "Version": 2.14,
   "Jobs": {
+    "async-sidecar": {
+      "2.12": {
```
Comment (nickboldt) [inline codeready-workspaces-devfileregistry/job-config.json:6]: remove 2.12 and 2.13 as these images don't exist in those releases.

File: codeready-workspaces-async-sidecar/dockerfiles/storage/ubi.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+# 
+
+# https://access.redhat.com/containers/?tab=tags#/registry.access.redhat.com/ubi8-minimal
+FROM registry.redhat.io/ubi8/ubi-minimal:8.3-230
+
+ADD content_sets_centos8.repo /etc/yum.repos.d/
```
Comment (nickboldt) [inline codeready-workspaces-async-sidecar/dockerfiles/storage/ubi.Dockerfile:12]: .repo files should be excluded from midstream and downstream -- Brew will choke on these

File: codeready-workspaces-async-sidecar/dockerfiles/storage/ubi.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2019-2020 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+# 
+
+# https://access.redhat.com/containers/?tab=tags#/registry.access.redhat.com/ubi8-minimal
+FROM registry.redhat.io/ubi8/ubi-minimal:8.3-230
```
Comment (nickboldt) [inline codeready-workspaces-async-sidecar/dockerfiles/storage/ubi.Dockerfile:10]: update to 8.5

File: crw-jenkins/jobs/CRW_CI/crw-async-sidecar.groovy [hunk 4]
```diff
@@ -0,0 +1,100 @@
+import groovy.json.JsonSlurper
```
Comment (nickboldt) [inline crw-jenkins/jobs/CRW_CI/crw-async-sidecar.groovy:1]: move this file to gitlab MR

File: codeready-workspaces-devfileregistry/job-config.json (hunk 1 above)
Comment (nickboldt) [reply codeready-workspaces-devfileregistry/job-config.json:6]: this is a local copy for the purposes of the devfile reg builds.

source file is https://github.com/redhat-developer/codeready-workspaces/blob/crw-2-rhel-8/dependencies/job-config.json

File: codeready-workspaces-async-sidecar/release.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+# Copyright (c) 2012-2019 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+
+#include common scripts
+. ./common.sh
```
Comment (nickboldt) [inline codeready-workspaces-async-sidecar/release.sh:12]: exclude the upstream release files like this one. no need in midstream/down.

File: codeready-workspaces-async-sidecar/dockerfiles/agent/ubi.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ENV USER=user \
+    UID=12345 \
+    GROUP=group \
+    GID=23456
+
+ADD content_sets_centos8.repo /etc/yum.repos.d/
+
+#cron task not work in openshift in case https://github.com/gliderlabs/docker-alpine/issues/381
+#so will used  supercronic https://github.com/aptible/supercronic
+ENV SUPERCRONIC_URL=https://github.com/aptible/supercronic/releases/download/v0.1.9/supercronic-linux-amd64 \
```
Comment (nickboldt) [inline codeready-workspaces-async-sidecar/dockerfiles/agent/ubi.Dockerfile:20]: will need to pre-fetch this and commit it as a lookaside cached binary in pkgs.devel -- curl is not allowed in brew.

so you'll want to look at a get-sources script like https://github.com/redhat-developer/codeready-workspaces-images/blob/crw-2-rhel-8/codeready-workspaces-plugin-java11/get-sources.sh#L73-L74

File: codeready-workspaces-async-storage-sidecar/build/scripts/sync.sh [hunk 7]
```diff
@@ -79,7 +79,6 @@ find ${TARGETDIR}/ -name "*.sh" -exec chmod +x {} \;
 # transform build/*/Dockerfile -> Dockerfile
 sed ${SOURCEDIR}/dockerfiles/${UPSTM_IMAGE}/ubi.Dockerfile -r \
   `# Replace ubi8 with rhel8 version` \
-  -e "s#ubi8/go-toolset#rhel8/go-toolset#g" \
```
Comment (nickboldt) [inline codeready-workspaces-async-storage-sidecar/build/scripts/sync.sh]: You need all three of these lines to switch to the version of go-toolset in OSBS registry.

File: codeready-workspaces-async-storage-sidecar/build/scripts/sync.sh [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+rsync -azrlt --checksum --exclude-from /tmp/rsync-excludes --delete ${SOURCEDIR}/ ${TARGETDIR}/
+rm -f /tmp/rsync-excludes
+
+# ensure shell scripts are executable
+find ${TARGETDIR}/ -name "*.sh" -exec chmod +x {} \;
+
+# transform build/*/Dockerfile -> Dockerfile
+sed ${SOURCEDIR}/dockerfiles/${UPSTM_IMAGE}/ubi.Dockerfile -r \
+  `# Replace ubi8 with rhel8 version` \
+  -e "s#FROM registry.redhat.io/#FROM #g" \
```
Comment (nickboldt) [inline codeready-workspaces-async-storage-sidecar/build/scripts/sync.sh:82]: need to add ` -e "s#ubi8/go-toolset#rhel8/go-toolset#g" \` if transforming from go-toolset image. 

Simpler transform could be this:

```
 `# Remove registry so build works in Brew` \
 -e "s#FROM (registry.access.redhat.com|registry.redhat.io)/#FROM #g" \
 `# Replace go-toolset ubi8 with rhel8 version` \
 -e "s#ubi8/go-toolset#rhel8/go-toolset#g" \
```

File: codeready-workspaces-async-storage-sidecar/build.sh [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        exit 0
+        ;;
+    *) printf "illegal option: -%s\n" "$key"
+       echo "$usage"
+       exit 1
+       ;;
+    esac
+done
+
+compile;
```
Comment (nickboldt) [inline codeready-workspaces-async-storage-sidecar/build.sh:56]: were you going to submit a PR for upstream that provides a 2-stage dockerfile that does the compile (builder stage) and then then dockerBuild (runtime stage) ?

File: codeready-workspaces-async-storage-sidecar/common.sh [hunk 10]
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
```
Comment (nickboldt) [inline codeready-workspaces-async-storage-sidecar/common.sh:1]: this file is not useful in downstream; if we can run the compilation `GOOS=linux go build -o ./dockerfiles/agent/scripts/watcher ./watcher/watcher.go` in a docker build, then most of this script becomes only useful for GHA releases to quay.

Comment (mkuznyetsov) [reply codeready-workspaces-async-storage-sidecar/common.sh:1]: added to exclusion

File: codeready-workspaces-async-storage-sidecar/build/scripts/sync.sh (hunk 8 above)
Comment (mkuznyetsov) [reply codeready-workspaces-async-storage-sidecar/build/scripts/sync.sh]: thx

File: codeready-workspaces-devfileregistry/job-config.json (hunk 1 above)
Comment (mkuznyetsov) [reply codeready-workspaces-devfileregistry/job-config.json]: so these changes gotta be removed from dev.reg. file?

File: codeready-workspaces-async-sidecar/dockerfiles/storage/ubi.Dockerfile (hunk 2 above)
Comment (mkuznyetsov) [reply codeready-workspaces-async-sidecar/dockerfiles/storage/ubi.Dockerfile]: got it, thanks

File: codeready-workspaces-async-storage-sidecar/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,91 @@
+# build go content
+FROM rhel8/go-toolset:1.16.7-5 as builder
```
Comment (nickboldt) [inline codeready-workspaces-async-storage-sidecar/Dockerfile:2]: needs the URL from which to pull updated base images 
```
# https://access.redhat.com/containers/?tab=tags#/registry.access.redhat.com/rhel8/go-toolset
```

File: codeready-workspaces-pluginbroker-artifacts/build/artifacts/Dockerfile [hunk 12]
```diff
@@ -15,7 +15,7 @@ FROM registry.access.redhat.com/ubi8/go-toolset:1.16.7-5 as builder
 USER root
 WORKDIR /build/che-plugin-broker/brokers/artifacts/cmd/
 COPY . /build/che-plugin-broker/
-RUN adduser appuser && \
+RUN adduser appusger && \
```
Comment (nickboldt) [inline codeready-workspaces-pluginbroker-artifacts/build/artifacts/Dockerfile:18]: `appusger` ?

File: codeready-workspaces-devfileregistry/job-config.json [hunk 13]
```diff
@@ -2,6 +2,22 @@
   "Purpose": "To set configurations of jenkins jobs including upstream branches, dependencies and disabled status",
   "Version": 2.14,
   "Jobs": {
+    "async-storage-sidecar": {
```
Comment (nickboldt) [inline codeready-workspaces-devfileregistry/job-config.json:5]: this is already merged, including adding 2.15 values. if merge fails, you can remove this

File: codeready-workspaces-async-storage-sidecar/dockerfiles/sidecar/ubi.Dockerfile [hunk 14]
```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2019-2020 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+
+# https://access.redhat.com/containers/?tab=tags#/registry.access.redhat.com/ubi8-minimal
+FROM registry.redhat.io/ubi8/ubi-minimal:8.3-230
```
Comment (nickboldt) [inline codeready-workspaces-async-storage-sidecar/dockerfiles/sidecar/ubi.Dockerfile:9]: this should be 8.5 in upstream and downstream too. Please submit a PR for upstream.

File: codeready-workspaces-devfileregistry/job-config.json (hunk 1 above)
Comment (nickboldt) [reply codeready-workspaces-devfileregistry/job-config.json]: no, that file should be pulled with each build.. I hope.

File: codeready-workspaces-async-storage-sidecar/dockerfiles/server/ubi.Dockerfile [hunk 15]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2019-2020 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+# 
+
+# https://access.redhat.com/containers/?tab=tags#/registry.access.redhat.com/ubi8-minimal
+FROM registry.redhat.io/ubi8/ubi-minimal:8.3-230
```
Comment (nickboldt) [inline codeready-workspaces-async-storage-sidecar/dockerfiles/server/ubi.Dockerfile:10]: please exclude dockerfile/server/* files from the sidecar build as they're not required.

File: codeready-workspaces-async-storage-sidecar/build/scripts/sync.sh [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+.git/
+.gitattributes
+build/scripts/sync.sh
+/container.yaml
+/content_sets.*
+/cvp.yml
+/cvp-owners.yml
+get-source*.sh
+tests/basic-test.yaml
+sources
```
Comment (nickboldt) [inline codeready-workspaces-async-storage-sidecar/build/scripts/sync.sh:63]: please exclude dockerfiles/server/* files as we don't need them in sidecar folder in midstream.

**Pull request merged**